### ISSUE TITLE
Cleanup Emscripten canvas target selector string stuff.

### ISSUE THREAD

**Issue by floooh** [floooh/sokol#1154]
See: https://github.com/floooh/sokol/issues/407#issuecomment-2480549987

TL;DR:

- clean up the confusion between DOM element id and selector strings
- change `sapp_desc.html5_canvas_name` to `sapp_desc.html5_canvas_selector`
- change the default from `canvas` to `#canvas`
- remove the special `#` prefixing
- in sokol_app.h JS code, use Emscripten's `findCanvasEventTarget()` and `findEventTarget()` instead of `getElementById()` or `querySelector()` so that the behaviour is identical with the Emscripten runtime functions.
- remove the two redundant getElementById in drag'n'drop init/shutdown code

**Comment by konsumer** [comment]
This path is still using strings (with `querySelector`.)

I recommend passing canvas dom-object (with a selector on `#canvas`/`sapp_desc.html5_canvas_selector` by default, for backwards compatibility.) I think emscripten might even set it up for you, if you just use `Module.canvas`. You can see an example in [glfw](https://github.com/emscripten-core/emscripten/blob/079e2660071d5ae7d45d8e04efc09dd7e67456dd/src/library_glfw.js#L1114)

This way I can override the canvas object, even if there is no document-selector (like in the case of a web-component.)

**Comment by floooh** [comment]
You can't pass a JS object across the WASM and C boundary, and you also can't pass a JS object into the `emscripten_*` functions.

**Comment by floooh** [comment]
PS: e.g. see here: https://emscripten.org/docs/tools_reference/settings_reference.html#disable-deprecated-find-event-target-behavior

`Module['canvas']` is deprecated but still works for backward compatibility.

**Comment by konsumer** [comment]
While I see that it's marked deprecated there, for my use case it really makes it less useful. I wonder what usecase that improves things for (seems like less control, with no payoff.) Still, it seems like a quick check once if canvas is a string (css selector) or dom element is the most consistent way to handle it, looking at other libs, even if it is deprecated behaviour. I could PR for both (so it handles new case and this deprecated thing thAt works everywhere else, and web components rely on.)

**Comment by konsumer** [comment]
Like maybe this is just that I don't know how to do it better, but there are situations where `document` does not point to the DOM that the canvas is in (like in case of shadow-dom) so `document.querySelector('#whatever')` doesn't work. 

I made [raylib-wasm](https://github.com/konsumer/raylib-wasm) and so I just followed what I did there, for example [this codepen](https://codepen.io/konsumer/pen/NWEdxZ) uses a web-component to make it easy to tie the draw-loop, instance & user-code to a single canvas. Is there a different way I should do that with sokol?

**Comment by konsumer** [comment]
I think maybe you are misunderstanding what I propose. I will make an example, using a branch.

**Comment by konsumer** [comment]
I think the "full solution" is to setup `Module.canvas` in `_sapp_init_state`, and remove all other direct-use of `_sapp.html5_canvas_selector`.

- [example](http://konsumer.js.org/sokol-canvas-example-1154/)
- [code](https://github.com/konsumer/sokol-canvas-example-1154)

I marked my changes with `DKDEMO` comments, but if this seems like a sensible path, I can PR and integrate it all much better. The main idea is I grab `Module.canvas`, in `_sapp_init_state' first, then later in `_sapp_emsc_webgl_init`. After that, any use of the current context (emscripten has wrapping around this) should work fine, and also, you have `Module.canvas` that you can use in any EM_JS/EM_ASM code, if needed. If I remember right, there are also some js-side helpers for integrating with emscripten C-side canvas/ctx helper-things, but I didn't bother figuring it all out, since it's simple enough to just do it in JS, directly. This supports the old behavior, as well as the new proposed behavior, and I made a demo that uses old, as well as web-components that use new.

For this quick-demo, I just did it for basic graphics (didn't touch any input handlers or anything.) This is all easy ebough to add (can be done in `EM_JS`, or using the emscripten html5 C wrapper things, which themselves use `EM_JS`) but it took me a while to just get a basic demo working (not a prob with emscripten API, more like just getting sokol working in build-env and stuff) and I need to get back to working on my game-engine.

**Comment by floooh** [comment]
Thanks for taking the time to create an example, much appreciate, and I will set aside some time to properly understand it.

I also asked around on the Emscripten Discord what the state of the special `Module['canvas']` handling in Emscripten is.

One thing I'm seeing is that you removed the `emscripten_get_element_css_size` and `emscripten_set_canvas_element_size` calls with your own versions which directly modify the JS canvas object. Is the reason for this that those functions stopped working?

OTH the calls to `emscripten_get_element_css_size` and `emscripten_set_canvas_element_size` in the `_sapp_emsc_run()` function are still there, I don't understand why the function is replaced in one place and not the other.

Also there's all those event handlers here:

https://github.com/konsumer/sokol-canvas-example-1154/blob/e72fb5488ee105a0930c35121ad04832a81fe696/sokol/sokol_app.h#L5895-L5911

...I wonder if those still work.

Also a question: wouldn't the problem of multiple canvases on the same page also be solved by giving each canvas its unique id and then give each demo that id-string in the `sokol_main()` function? This assumes that the `.html5_canvas_selector` is passed through as is, and the special `#` prefixing is removed.

**Comment by konsumer** [comment]
> OTH the calls to emscripten_get_element_css_size and emscripten_set_canvas_element_size in the _sapp_emsc_run() function are still there, I don't understand why the function is replaced in one place and not the other.

maybe. That was a quick hack to move away from a selector, but as I said `!canvas` should work, too. Overall, it seems like the lightest way to keep all the code the same, except for the init entry-point (where `Module.canvas` is setup.)

> ...I wonder if those still work.

Not in my demo, I didn't touch them, but if the selector is set to `!canvas` it should. We did similar at [pntr_app](https://github.com/RobLoach/pntr_app/blob/master/include/pntr_app_web.h)

> Also a question: wouldn't the problem of multiple canvases on the same page also be solved by giving each canvas its unique id and then give each demo that id-string in the sokol_main() function? This assumes that the .html5_canvas_selector is passed through as is, and the special # prefixing is removed.

Maybe, if you are not using web-components, but in my example, as I said, the canvas in web-component is not even in `document`, so you can't grab it with an id/selector. Additionally, since the way it's setup is that the selector is hard-coded in the wasm-code, I could not run 2 instances of the same wasm in the same page, anyway.

**Comment by floooh** [comment]
(I added more stuff to my previous comment while you wrote your reply).

> Maybe, if you are not using web-components

...if that means that the current Emscripten runtime functions which take a target-string parameter is no compatible with web-components, then I would rather prefer to not use web components instead of not being able to use those Emscripten functions tbh.

**Comment by floooh** [comment]
I'll see if I can make some time to tinker around with the ideas in this thread (with the main target to easily enable multiple WebGL canvases on the same page).

In any case, thanks for the input so far :)

**Comment by konsumer** [comment]
> I wonder if it is enough to initialize Module['canvas'] outside WASM to make all the emscripten_* functions work. 

Yeh, that is what I am proposing, now. I forgot about `specialHTMLTargets`, which works great for this specific usecase in pntr_app, but my original point was "it's all js under that C code" like you can rewire anything in `EM_JS`. I think with `specialHTMLTargets` everything will basically work the same, though, with 1 lil js hook at the init.

> ...if that means that the current Emscripten runtime functions which take a target-string parameter are not compatible with web-components, then I would rather prefer to not use web components instead of not being able to use those Emscripten functions tbh

There are several parts to your question, I will elaborate:

- You can't use an id or query-selector at all, in a practical sense, and target multiple canvases. Even without web-components, I would need to compile several copies of my game to run on the same page because the selector is compiled in the C code. if I had 2 copies of that, then I would need to compile 2 different versions, each targeting a different selector, and I would have to carefully make sure the canvases had the right IDs, in any page they are used in.
- there is nothing specific to web-components here. Think of the ask as "I would rather pass my own canvas instead of a string, so I can tell the emscripten runtime which canvas to use in a better way." This is 100% doable, and pretty simple with `!canvas` trick. All users would benefit, if they want to be able to connect their sokol wasm up to different/separate canvases.
- the Emscripten functions are just wrappers around `EM_JS`. it's all the same thing. I think it's clumsy (often the wrapper functions don't work quite right/like docs, or don't have the same configurability as the underlying js, which they actually use themselves.) I tend to write stuff directly with js in general, using standard web APIs, but we could make it more in C-space, if you like that better. Again, using the `!canvas` trick, pretty much everything works exactly the same, and barely any code would change (like 2 lines in the whole file.)
- no offense, but it's not really up to you to "use web components" in my project, like I am already using them. I can not use sokol, or fork it and add a couple lines to the header to make it work, if I want to switch to sokol, but that all seems so silly, because it really does support more options, without taking anything away. It seems like a win-win to just add it, so everyone can use sokol as they do now, but also have a couple more options.

> PS: the specialHTMLTargets thing looks like a good solution, unless this is also part of the deprecated behaviour... hmm.

Again, the deprecated part, as I understand it, is that emscripten used to automatically wire itself to `#canvas` via `Module.canvas`. Supporting both things doesn't take anything away or tie sokol to some old, bad behavior. Currently sokol is more like the deprecated behavior, anyway (it binds itself to `#canvas` by default.)

**Comment by floooh** [comment]
> the Emscripten functions are just wrappers around EM_JS. it's all the same thing.

...that's true, but the Emscripten shim functions also sometimes have convenient workarounds for web API problems I don't necessarily want to deal with myself. In general I prefer the Emscripten functions over my own EM_JS functions, unless there's a really good reason not to use them.

**Comment by konsumer** [comment]
> ...that's true, but the Emscripten shim functions also sometimes have convenient workarounds for web API problems I don't necessarily want to deal with myself. In general I prefer the Emscripten functions over my own EM_JS functions, unless there's a really good reason not to use them.

I disagree, but it's a matter of taste I think, and I am fine with doing it either way. Again, with `!canvas` it's a bit of a moot point, since almost all of the code will be exactly the same (doing everything as you do, now.)

**Comment by floooh** [comment]
> but it's not really up to you to "use web components" in my project

Well if it means kicking out all `emscripten_*` functions from the sokol header then that's a pretty good reason because it means increased maintenance burden. It looks like the `specialHTMLTarget[]` might solve that issue though.

**Comment by konsumer** [comment]
> Well if it means kicking out all emscripten_* functions 
it doesn't. that is what I'm saying.

**Comment by konsumer** [comment]
I am happy to make another demo that uses `!canvas`/`specialHTMLTarget` and changes absolutely as little sokol code as possible. I wasn't really aiming for that before, because I was saving time by not having to look things up (I am already pretty good with the js apis, or they are at least familiar) but you have done all the C work already, and it really just needs like 1 dumb little `EM_JS` function to tie `!canvas` to a sensible thing.

**Comment by floooh** [comment]
If you try that, you'd might at least want to replace this getElementById() line:

https://github.com/floooh/sokol/blob/2c6fc7470e9b9121a178e6e68c55f2f06fac4647/sokol_app.h#L5020

Ideally this would call Emscripten's `findCanvasEventTarget()` function, hopefully that's available anywhere.

You might also need to add `$findCanvasEventTarget` to the EM_JS_DEPS() here:

https://github.com/floooh/sokol/blob/2c6fc7470e9b9121a178e6e68c55f2f06fac4647/sokol_app.h#L4783

**Comment by floooh** [comment]
PS: also look for places which skip the first character in like `_sapp.html5_canvas_selector[1]`, e.g. change this line:

https://github.com/floooh/sokol/blob/2c6fc7470e9b9121a178e6e68c55f2f06fac4647/sokol_app.h#L5934

...to `sapp_js_init(_sapp.html5_canvas_selector);`

...I think that should take care of most things (drag'n'drop won't work but that's not critical for the experiment).

**Comment by konsumer** [comment]
yep, I think there are sort of 2 parts to this, like:

- switch out all the direct id stuff to use query-selector
- allow user to add canvas option instead of strings, and use `!canvas` as selector, if that option was provided

I think the first is only in a few places, since everything mostly uses selectors anyway.

**Comment by konsumer** [comment]
> drag'n'drop won't work but that's not critical for the experiment

I think if it all uses query-selector, it should work fine, right? Things in js-space can use `Module.canvas` and things in C-space can use `!canvas` as the selector.

**Comment by konsumer** [comment]
Ok, I made more minimal changes, swapping out `!canvas` where appropriate. I also swapped out `Module.sapp_emsc_target`, since it's a duplication of `Module.canvas`, anyway.

> You might also need to add $findCanvasEventTarget to the EM_JS_DEPS()

I ended up just doing this, since now we have a link to it. You actually already had `Module.sapp_emsc_target`, so all that stuff was a bit redundant anyway:

```js
const canvas = Module.canvas;
```

I didn't test DnD, but it should work, too.

**Comment by floooh** [comment]
Thanks! Looks a bit different than I imagined. I might do a counterproposal, which we can compare then ;)

**Comment by konsumer** [comment]
> Thanks! Looks a bit different than I imagined. I might do a counterproposal, which we can compare then ;)

There are basically 2 ways to target the "screen canvas", depending on if you are in JS or C, in my solution, and I am not sure how that could be improved upon:

- if in C (which uses those wrapper functions) you can use `!canvas` as the selector, so not need for a selector param. It's already wired up to `Module.canvas`
- If in JS (`EM_JS`, etc) you can use `Module.canvas`, it already has the canvas.

I will suggest an improvement to my path, though: any sokol functions themselves do not need a string-selector arg, anymore. Like after init, id and selector should just not be touched/passed, ever. I tried to keep everything as it is, which means they all have unused selector-string params.

I think my PR could be merged for an immediate improvement (supports `canvas` option, which enables multiple canvases on a page for same wasm) and that code-cleanup could be done later. Since I am not actually using sokol in anything, it would be more appropriate for you or someone that works on/uses sokol to change it.

**Comment by floooh** [comment]
Here's my 'counter-proposal':

https://github.com/floooh/sokol/pull/1159

Canvases that can't be looked up via `document.querySelector()` would need to be registered via the `specialHTMLTargets` map. So far I did some simple tests with different canvas ids and also using the specialHTMLTargets methods, next I'll try to integrate that with your 'multiple canvas' example.

I also tested the above PR with samples that use the input event handlers and drag'n'drop stuff.
